internals/overlord/checkstate/manager.go

```diff
@@ -48,42 +75,53 @@ func (m *CheckManager) NotifyCheckFailed(f FailureFunc) {
 
 // PlanChanged handles updates to the plan (server configuration),
 // stopping the previous checks and starting the new ones as required.
-func (m *CheckManager) PlanChanged(p *plan.Plan) {
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
-
-	logger.Debugf("Configuring check manager (stopping %d, starting %d)",
-		len(m.checks), len(p.Checks))
-
-	// First stop existing checks.
-	for _, check := range m.checks {
-		check.cancel()
-	}
-	// Wait for all context cancellations to propagate and allow
-	// each goroutine to cleanly exit.
-	m.wg.Wait()
-
-	// Set the size of the next wait group
-	m.wg.Add(len(p.Checks))
-
-	// Then configure and start new checks.
-	checks := make(map[string]*checkData, len(p.Checks))
-	for name, config := range p.Checks {
-		ctx, cancel := context.WithCancel(context.Background())
-		check := &checkData{
-			config:  config,
-			checker: newChecker(config, p),
-			ctx:     ctx,
-			cancel:  cancel,
-			action:  m.callFailureHandlers,
+func (m *CheckManager) PlanChanged(newPlan *plan.Plan) {
```

Comment:
Q: Do you think there could be some benefit from decoupling plan change induced work from the ```PlanChanged``` handler itself? I like the idea of never having the ability for a notification consumer to lock up or delay the writer, especially if its a global writer attending to all managers in the system.

So I deal with the repercussions of the new plan within my own event routine. I also use the non-blocking write pattern we briefly discussed, allowing the consumer no take another plan change unless its ready do so, without an impact on the writer. It could miss an incremental plan change in the process, but it would always get the latest change at least.

```
// PlanChanged is called by the Plan Manager when the plan is changed.
func (m *NetManager) PlanChanged(plan *plan.Plan) {
	// This is a single writer non-block pattern using a single entry
	// buffered channel. Remove entry if it exist, and [re-]insert.
	select {
	case <-m.planChanged:
	default:
	}
	m.planChanged <- plan
}
```

Stealing your PR for feedback on my PR ;)

Reply:
I think it's an interesting and useful idea. And I think if `PlanChanged` were doing something even slightly long-running I'd use it. However, I don't think the plan will change very often, and whatever goroutine that needs to update the `CheckManager` will still need to take the locks and whatnot, so I'm not sure it'll help much here.

Reply:
Per voice discussion, the problem with that method is if two people are concurrently calling the API to append/combine a layer, the second one may block. The above technique is only safe if you can guarantee that there's only one concurrent caller of `PlanChanged`, which isn't true for us.

I've gone with standard locking, though I've rearranged it quite a bit.